akka-stream/src/main/scala/akka/stream/scaladsl/RetryFlow.scala

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.scaladsl
+
+import akka.pattern.BackoffSupervisor
+import akka.stream.scaladsl.RetryFlowCoordinator.InternalState
+import akka.stream.stage.{ GraphStage, GraphStageLogic, InHandler, OutHandler, TimerGraphStageLogic }
+import akka.stream.{ Attributes, BidiShape, FlowShape, Inlet, Outlet }
+
+import scala.collection.immutable
+import scala.concurrent.duration._
+import scala.util.Try
+
+object RetryFlow {
+
+  /**
+   * Allows retrying individual elements in the stream with an exponential backoff.
+   *
+   * The retry condition is controlled by the `retryWith` partial function. It takes an output element of the wrapped
+   * flow and should return one or more requests to be retried. For example:
+   *
+   * case Failure(_) => Some(List(..., ..., ...)) - for every failed response will issue three requests to retry
+   *
+   * case Failure(_) => Some(Nil) - every failed response will be ignored
+   *
+   * case Failure(_) => None - every failed response will be propagated downstream
+   *
+   * case Success(_) => Some(List(...)) - for every successful response a single retry will be issued
+   *
+   * A successful or failed response will be propagated downstream if it is not matched by the `retryFlow` function.
+   *
+   * If a successful response is matched and issued a retry, the response is still propagated downstream.
+   *
+   * The implementation of the RetryFlow assumes that `flow` follows one-in-one-out element semantics.
+   *
+   * The wrapped `flow` and `retryWith` takes an additional `State` parameter which can be used to correlate a request
+   * with a response.
+   *
+   * Backoff state is tracked separately per-element coming into the wrapped `flow`.
+   *
+   * @param parallelism controls the number of in-flight requests in the wrapped flow
+   * @param minBackoff minimum duration to backoff between issuing retries
+   * @param maxBackoff maximum duration to backoff between issuing retries
+   * @param randomFactor adds jitter to the retry delay. Use 0 for no jitter
+   * @param flow a flow to retry elements from
+   * @param retryWith retry condition decision partial function
+   */
+  def withBackoff[In, Out, State, Mat](
```

Comment:
I think we should mark it APIMayChange to not set the factory signature in stone right away, that parameter list sure is a mouthful.